### PR TITLE
ENH: Adding IncrementalRidge support into sklearnex

### PR DESCRIPTION
**Pull request by DDJHB**
### Description 
This pr introduces Incremental Ridge to sklearnex.

Fixes # - _issue number(s) if exists_

- [x] I have reviewed my changes thoroughly before submitting this pull request.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation to reflect the changes, if necessary (updated in # - _add PR number_)
- [x] The unit tests pass successfully.
- [x] I have run it locally and tested the changes extensively.
- [x] I have resolved any merge conflicts that might occur with the base branch.
- [x] Git commit message contains an appropriate signed-off-by string _(see [CONTRIBUTING.md](https://github.com/intel/scikit-learn-intelex/blob/main/CONTRIBUTING.md#pull-requests) for details)_
- [ ] I have added a respective label(s) to PR if I have a permission for that.

### CODE REVIEW

File: sklearnex/preview/linear_model/tests/test_incremental_ridge.py [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+    from onedal.tests.utils._dataframes_support import (
+        _as_numpy,
+        _convert_to_dataframe,
+        get_dataframes_and_queues,
+    )
+    from sklearnex.preview.linear_model import IncrementalRidge
+
+    @pytest.mark.parametrize("dataframe,queue", get_dataframes_and_queues())
+    @pytest.mark.parametrize("batch_size", [10, 100, 1000])
+    @pytest.mark.parametrize("alpha", [0.1, 0.5, 1.0])
```
Comment (olegkkruglov) [inline sklearnex/preview/linear_model/tests/test_incremental_ridge.py:34]: please add fit_intercept parametrizing

File: sklearnex/preview/linear_model/tests/test_incremental_ridge.py [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+            fit_intercept=False, alpha=alpha, batch_size=batch_size
+        )
+        inc_ridge.fit(X_c, y_c)
+
+        lambda_identity = alpha * np.eye(X.shape[1])
+        inverse_term = np.linalg.inv(np.dot(X.T, X) + lambda_identity)
+        xt_y = np.dot(X.T, y)
+        coefficients_manual = np.dot(inverse_term, xt_y)
+
+        assert_allclose(inc_ridge.coef_, coefficients_manual, rtol=1e-6, atol=1e-6)
```
Comment (olegkkruglov) [inline sklearnex/preview/linear_model/tests/test_incremental_ridge.py:52]: please add predict test which check that the value is expected, not only shape

File: onedal/linear_model/incremental_linear_model.py [hunk 3]
```diff
@@ -144,3 +144,113 @@ def finalize_fit(self, queue=None):
...
+            If not None, use this queue for computations.
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        module = self._get_backend("linear_model", "regression")
+
+        if not hasattr(self, "_policy"):
+            self._policy = self._get_policy(queue, X)
```
Comment (olegkkruglov) [inline onedal/linear_model/incremental_linear_model.py:207]: This part will be changed in other incremental algorithms after merging incremental distributed PRs. I think it is better to align here as well. please check, for example, https://github.com/intel/scikit-learn-intelex/pull/1941 and change onedal/linear_model/incremental_linear_model.py and sklearnex/preview/linear_model/incremental_ridge.py accordingly.

**Comment by ethanglaser** [comment]
General comment: not sure this needs to go into preview. A bit of a unique situation since batch ridge regression is currently in preview, but considering there is no stock scikit-learn IncrementalRidge estimator it can probably go straight into sklearnex/linear_model

File: sklearnex/preview/linear_model/tests/test_incremental_ridge.py (hunk 2 above)
Comment (DDJHB) [reply sklearnex/preview/linear_model/tests/test_incremental_ridge.py:52]: nice point, i will change predict test separately to include this

**Comment by icfaust** [comment]
Private CI Green with unrelated failures.

File: onedal/linear_model/incremental_linear_model.py [hunk 4]
```diff
@@ -144,3 +144,118 @@ def finalize_fit(self, queue=None):
...
+
+        packed_coefficients = from_table(result.model.packed_coefficients)
+        self.coef_, self.intercept_ = (
+            packed_coefficients[:, 1:],
+            packed_coefficients[:, 0],
+        )
+
+        if self.coef_.shape[0] == 1 and self._y_ndim_1:
+            self.coef_ = self.coef_.ravel()
+            self.intercept_ = self.intercept_[0]
```
Comment (icfaust) [inline onedal/linear_model/incremental_linear_model.py:259]: I know this code came from the IncrementalLinearRegression, but _y_ndim_1 seems somehow unnecessary.  Can't the dimensions of y be queried from the intercept or coefficients? i.e. cant we just use np.squeeze on the result?

File: onedal/linear_model/tests/test_incremental_ridge_regression.py [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+        for i in range(2):
+            model.partial_fit(X_train_split[i], y_train_split[i], queue=queue)
+        model.finalize_fit()
+        y_pred = model.predict(X_test, queue=queue)
+        assert_allclose(mean_squared_error(y_test, y_pred), 2388.775, rtol=1e-5)
+
+    @pytest.mark.parametrize("queue", get_queues())
+    @pytest.mark.parametrize("dtype", [np.float32, np.float64])
+    @pytest.mark.skip(reason="pickling not implemented for oneDAL entities")
+    def test_pickle(queue, dtype):
```
Comment (icfaust) [inline onedal/linear_model/tests/test_incremental_ridge_regression.py:50]: Just make sure there is a ticket or note in another ticket to remove this skip when serialization is implemented for IncrementalRidge

File: sklearnex/linear_model/incremental_ridge.py [hunk 6]
```diff
@@ -0,0 +1,418 @@
...
+from onedal.linear_model import IncrementalRidge as onedal_IncrementalRidge
+
+from .._device_offload import dispatch, wrap_output_data
+from .._utils import PatchingConditionsChain
+
+
+@control_n_jobs(
+    decorated_methods=["fit", "partial_fit", "predict", "_onedal_finalize_fit"]
+)
+class IncrementalRidge(MultiOutputMixin, BaseEstimator, RegressorMixin):
```
Comment (icfaust) [inline sklearnex/linear_model/incremental_ridge.py:41]:
```suggestion
class IncrementalRidge(MultiOutputMixin, RegressorMixin, BaseEstimator):
```

BaseEstimator is generally latest in the MRO for sklearn estimators.  I'm not sure what may overwrite the RegressorMixin or MultiOutputMixin, but just to be sure and to match convention.

File: sklearnex/linear_model/incremental_ridge.py [hunk 7]
```diff
@@ -0,0 +1,418 @@
...
+                "onedal": self.__class__._onedal_partial_fit,
+                "sklearn": None,
+            },
+            X,
+            y,
+            check_input=check_input,
+        )
+        return self
+
+    def fit(self, X, y):
```
Comment (icfaust) [inline sklearnex/linear_model/incremental_ridge.py:272]: Maybe I'm opening pandora's box here but Ridge takes sample_weight https://github.com/scikit-learn/scikit-learn/blob/main/sklearn/linear_model/_ridge.py#L911 . This is also the case for preview Ridge.  Please add a keyword for conformance purposes and add an error if its not None. I know that may seem unnecessary, but generally trying to conform to related estimators is top priority.  You can let me know if you find that unneccessary.

**Review by icfaust — COMMENTED** [review]
Small changes, otherwise first round looks good.

File: onedal/linear_model/incremental_linear_model.py (hunk 4 above)
Comment (DDJHB) [reply onedal/linear_model/incremental_linear_model.py:259]: thanks, nice suggestion

File: sklearnex/dispatcher.py [hunk 8]
```diff
@@ -412,6 +413,19 @@ def get_patch_map_core(preview=False):
             ]
         ]
 
+        if daal_check_version((2024, "P", 600)):
```
[the remark this answers is not in the document]
Comment (DDJHB) [reply sklearnex/dispatcher.py:416]: adding IncrementalRidge to patching (not preview) without this check fails public ci, since it only works from onedal version 2024.6.0. there is one test for test_patching that takes estimators from patching and calls their methods, however it fails on backend since public ci uses 2024.5.0

File: onedal/linear_model/incremental_linear_model.py [hunk 9]
```diff
@@ -144,3 +144,114 @@ def finalize_fit(self, queue=None):
...
+            policy = self._get_policy(self._queue)
+        result = module.finalize_train(policy, self._params, self._partial_result)
+
+        self._onedal_model = result.model
+
+        packed_coefficients = from_table(result.model.packed_coefficients)
+        self.coef_, self.intercept_ = (
+            packed_coefficients[:, 1:].squeeze(),
+            packed_coefficients[:, 0].squeeze(),
+        )
```
[the remark this answers is not in the document]
Comment (DDJHB) [reply onedal/linear_model/incremental_linear_model.py:255]: yes, the 2 differences are: we do not use hyperparams for ridge yet and i have changed how final results are parsed thx to @icfaust [suggestion](https://github.com/intel/scikit-learn-intelex/pull/1957#discussion_r1708746789). let me know if you meant sth else

**Review by ethanglaser — APPROVED** [review]

**Review by icfaust — APPROVED** [review]
Small change on line 217 of onedal/linear_model/incremental_linear_model.py otherwise good to merge
